### PR TITLE
Add user registration functionality

### PR DESCRIPTION
**Pull request by chertiav**
- created the users entity;
- added an authentication controller;
- added RegistrationException and InstanceCreationException;
- added relevant DTOs;
- improved CustomGlobalExceptionHandler.

### CODE REVIEW

File: src/main/java/com/achdev/onlinebookstoreapp/dto/user/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    private String repeatPassword;
+
+    @NotBlank(message = "First name is mandatory")
+    private String firstName;
+
+    @NotBlank(message = "Last name is mandatory")
+    private String lastName;
+
+    @NotNull(message = "Shipping address cannot be null")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/achdev/onlinebookstoreapp/dto/user/UserRegistrationRequestDto.java]: follow the task description it is an optional field

File: src/main/java/com/achdev/onlinebookstoreapp/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserDto register(UserRegistrationRequestDto requestDto) throws RegistrationException {
+        String email = requestDto.getEmail();
+        if (userRepository.findByEmail(email).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/achdev/onlinebookstoreapp/service/impl/UserServiceImpl.java]: better use existByEmail method

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                  name: last_name
+                  type: VARCHAR(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: VARCHAR(255)
+
+
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]:
```suggestion
```
leave only one empty line

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes ;)

File: src/main/java/com/achdev/onlinebookstoreapp/dto/user/UserRegistrationRequestDto.java (hunk 1 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/dto/user/UserRegistrationRequestDto.java]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/service/impl/UserServiceImpl.java (hunk 2 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/service/impl/UserServiceImpl.java]: Ok

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml (hunk 3 above)
Comment (chertiav) [reply src/main/resources/db/changelog/changes/02-create-users-table.yaml]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/dto/errors/BaseErrorDto.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.achdev.onlinebookstoreapp.dto.errors;
+
+import java.time.LocalDateTime;
+import lombok.Data;
+
+@Data
+public abstract class BaseErrorDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/dto/errors/BaseErrorDto.java]: Suggestion
```
public record BookApiErrorResponse(
        HttpStatus status,
        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd-MM-yyyy HH:mm:ss")
        LocalDateTime timestamp,
        Object errorMessages
) {}
```

File: src/main/java/com/achdev/onlinebookstoreapp/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -26,30 +29,46 @@ protected ResponseEntity<Object> handleMethodArgumentNotValid(
...
-                new EntityNotFoundErrorResponseDto();
-        exceptionResponseDto.setTimestamp(LocalDateTime.now());
-        exceptionResponseDto.setStatus(HttpStatus.NOT_FOUND.getReasonPhrase());
-        exceptionResponseDto.setError("Entity not found");
-        exceptionResponseDto.setMessage(ex.getMessage() != null
-                ? ex.getMessage()
-                : "Object was not found");
-        exceptionResponseDto.setPath(request.getDescription(false));
-        return new ResponseEntity<>(exceptionResponseDto, HttpStatus.NOT_FOUND);
+    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/exception/CustomGlobalExceptionHandler.java:37]: Suggestion (you can use such approach for other exceptions) and remove other error dto that you created
```suggestion
    @ExceptionHandler(EntityNotFoundException.class)
    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {
        return buildResponseEntity(new BookApiErrorResponse(NOT_FOUND,
                LocalDateTime.now(), ex.getMessage()));
    }
```

File: src/main/java/com/achdev/onlinebookstoreapp/exception/InstanceCreationException.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.achdev.onlinebookstoreapp.exception;
+
+public class InstanceCreationException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/exception/InstanceCreationException.java]: Where did you got this exception?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/achdev/onlinebookstoreapp/exception/InstanceCreationException.java (hunk 6 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/exception/InstanceCreationException.java]: private <T> T createInstance(Class<T> clazz) {
        try {
            return clazz.getDeclaredConstructor().newInstance();
        } catch (Exception e) {
            throw new InstanceCreationException("Failed to create instance of class: "
                                                + clazz.getName());
        }
    }

File: src/main/java/com/achdev/onlinebookstoreapp/dto/errors/BaseErrorDto.java (hunk 4 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/dto/errors/BaseErrorDto.java]: Ок

File: src/main/java/com/achdev/onlinebookstoreapp/exception/CustomGlobalExceptionHandler.java (hunk 5 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/exception/CustomGlobalExceptionHandler.java:37]: ОК

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**